File: 競プロ典型90問/typical90/typical90_af/typical90_af.py
n = int(input())
a = [list(map(int, input().split())) for i in range(n)]
m = int(input())
friend = set()
for i in range(m):
  x, y = map(int, input().split())
  x, y = x - 1, y - 1
  friend.add((x, y))
  friend.add((y, x))

inf = 10 ** 18
dp = [[inf] * n for i in range(1 << n)]

for i in range(n):
  dp[1 << i][i] = min(dp[1 << i][i], a[i][0])

for s in range(1, (1 << n) - 1):
  t = format(s, 'b')
  j = t.count('1')

  for v in range(n):
    if not (s >> v) & 1:
      continue
    for u in range(n):
      if (s >> u) & 1:
        continue
      if (v, u) not in friend:
        ns = s | (1 << u)
        dp[ns][u] = min(dp[s][v] + a[u][j], dp[ns][u])

ans = min(dp[-1])
print(-1 if ans >= inf else ans)

File: 競プロ典型90問/typical90/typical90_af/typical90_af.cpp
#include <bits/stdc++.h>
#define all(x) (x).begin(), (x).end()
#define fix(i) fixed << setprecision(i)
#define fore(i, a) for(auto& i : a)
#define ll long long
#define nextp(v) next_permutation(all(v))
#define rep(i, a, b) for(int i = a; i < b; i++)
#define reps(i, a, b) for(int i = a; i <= b; i++)
#define rrep(i, a, b) for(int i = a; i > b; i--)
#define rreps(i, a, b) for(int i = a; i >= b; i--)
#define repv(i, a, b, c) for(int i = a; i < b; i += c)
#define repsv(i, a, b, c) for(int i = a; i <= b; i += c)
#define rrepv(i, a, b, c) for(int i = a; i > b; i -= c)
#define rrepsv(i, a, b, c) for(int i = a; i >= b; i -= c)
using namespace std;

int main() {
  cin.tie(nullptr);
  ios_base::sync_with_stdio(false);

  int n;
  cin >> n;
  vector<int> runner;
  rep(i, 0, n) runner.push_back(i);

  int a[n][n];
  rep(i, 0, n) rep(j, 0, n) cin >> a[i][j];

  int m;
  cin >> m;

  vector<int> bad[n];
  int x, y;
  while(cin >> x >> y) {
    bad[x - 1].push_back(y - 1);
    bad[y - 1].push_back(x - 1);
  }
  
  ll ans = 1e18;
  do {
    ll sum = 0;
    bool goal = true;

    rep(i, 0, runner.size() - 1) {
      if(count(all(bad[runner[i]]), runner[i + 1]) != 0) goal = false;
      sum += a[runner[i]][i];
    }
    sum += a[runner.back()][runner.size() - 1];

    if(goal) ans = min(ans, sum);
  } while(nextp(runner));

  cout << (ans == 1e18 ? -1 : ans);
  return 0;
}